### ISSUE TITLE
Conan Dependencies: Next Steps

### ISSUE THREAD

**Issue by jmarrec** [NREL/conan-openstudio-ruby#8]
Following #7 Next steps:

* [ ] Investigate why `with_libyaml=True` fails to link on Windows
* [x] Instead of always force-building some conan dependencies in OpenStudio/ConanInstall.cmake, find a better alternative.
    * currently on mac for eg websocket, boost_asio and readline are always built **each time you call cmake**
    * one is to upload a nrel/readline recipe that doesn't contain binary packages, so that the first time you build OpenStudio you must build it from source
* [ ] Investigate why conan-libffi doesn't succeed to link on Windows (I started to tweak the recipe here: https://github.com/jmarrec/conan-libffi/tree/testing/3.2.1)
* [ ] Remove the need for CMakeLists.txt: it works on Unix but having problems on windows due to libffi. Given the urgency of our deliverables I have put this effort on hold, but it's saved on https://github.com/jmarrec/conan-openstudio-ruby/pull/3
* [ ] Missing deps on windows
    * [ ] Provide readline via conan on MSVC
    * [ ] provide gdbm
* [x] Check if possible to enable static builds of conan-readline on macOS
* [ ] Check if needed to provide conan-gmp on Windows:
     * [ ] If so, check if possible to provide conan-gmp on Windows

**Comment by tijcolem** [comment]
- [x] Check if possible to enable static builds of conan-readline on macOS

For the item above,  I looked around and could not find options to build readline statically for osx.  Using the conan dylib readline libs is causing issues as openstudio cli has references to the libhistory.dylib and libreadline.dylib at runtime, so running the cli on another machine that does not have these, it crashes. 

e.g.
 `$ openstudio openstudio_version`
`dyld: Library not loaded: /Users/jenkins/.conan/data/readline/7.0/bincrafters/stable/package/988863d075519fe477ab5c0452ee71c84a94de8a/lib/libhistory.7.dylib`

If we can't build static readline, then I think we need to ship these dylibs with the cli.  Conan has a [guide](https://docs.conan.io/en/latest/howtos/manage_shared_libraries/rpaths.html) on how to modify rpath and import the dylibs.  I think this would work but maybe use this as last resort if we can't get static conan readline to work? Or we could use the brew version of readline that does have static. 

For now I think we should fall back to using the brew readline version that has static lib.

**Comment by jmarrec** [comment]
* [ ] use conan SWIG: A week or two ago a swig_installer package was published on bincrafters, with SWIG 4.0 (cf [bintray](https://bintray.com/beta/#/bincrafters/public-conan/swig_installer:bincrafters?tab=overview), [inclusion request](https://github.com/bincrafters/community/issues/610), [github repo](https://github.com/bincrafters/conan-swig_installer/tree/stable/4.0.0))
    * Not sure if SWIG 4.0 would create problems, given that we currently use 3.0.12, cf 4.0 release notes here: https://github.com/swig/swig/blob/master/RELEASENOTES#L10:L37
     * "C++11 STL containers added for Java, Python and Ruby." > this might clash with the custom stuff we had to do before.

**Comment by macumber** [comment]
I've been wanting to try out the SWIG 4 release.  I think we can try that out now once we have a good baseline of develop3 to compare against.  I'm not sure that anyone has tried the C# bindings in develop3 yet.

**Comment by jmarrec** [comment]
For conan-readline on mac as static, I tweaked the conan-readline recipe, and I have not seen any side effects of building static readline yet. cf https://github.com/bincrafters/conan-readline/pull/8

I then proceeded to get conan-openstudio-ruby to use my fork of readline static, and that builds and links correctly as well, and test_package works as before, https://github.com/NREL/conan-openstudio-ruby/pull/17

I will test a full OpenstudioApplication build now to check if that solves our homebrew dependency.
* I set up a branch of OpenStudio, https://github.com/NREL/OpenStudio/tree/test_static_readline_mac,  that uses that fork of conan-openstudio-ruby
* I then set up a branch of OpenStudioApplication also named test_static_readline_mac that tracks this OpenStudio branch test_static_readline_mac...
    * So CI should try to build this, and we can see how mac is doing and test the resulting package


So to recap: 
conan-readline > NREL/conan-openstudio-ruby:test_static_readline_mac > NREL/OpenStudio:test_static_readline_mac > NREL/OpenStudioApplication:test_static_readline_mac
